С-9-1. вынос запросов на уровень репозитория и добавление проверки в …

AleksandrTimonin opened the pull request:
…сервис временных отметок

AleksandrTimonin committed:
С-9-1. вынос запросов на уровень репозитория и добавление проверки в сервис временных отметок

sosulka7 reviewed (approved):
тока конфликт небольшой

itananina commented on core-service/src/main/java/com/sanjati/core/repositories/TimePointRepository.java:
```diff
@@ -2,11 +2,17 @@
 @Repository
 public interface TimePointRepository extends JpaRepository<TimePoint, Long>, JpaSpecificationExecutor<TimePoint> {
+   @Query("select count (tp) from TimePoint tp where tp.executorId = :executorId and tp.status = :status" )
+    public Integer checkByExecutorIdAndStatus(Long  executorId, TimePointStatus status);
```
jpa сама умеет каунты считать, тут можно не прописывать квери
Long countByExecutorIdAndStatus(Long executorId, TimePointStatus status);

itananina commented on core-service/src/main/java/com/sanjati/core/repositories/TimePointRepository.java:
```diff
@@ -2,11 +2,17 @@
+   @Query("select count (tp) from TimePoint tp where tp.executorId = :executorId and tp.status = :status" )
+    public Integer checkByExecutorIdAndStatus(Long  executorId, TimePointStatus status);
+   @Query("select count (tp) from TimePoint tp where tp.task.id = :taskId")
+    public Integer checkCountTimePoints(Long taskId);
```
а тут не сможет тк в TimePoint нет поля taskId, @Query оставляем

itananina replied:
тока подпиши пж поконкретнее типа countByTaskId (слово таймпоинт можно не упоминать тк каунтим через TimePointRepository)

itananina commented on core-service/src/main/java/com/sanjati/core/repositories/TaskRepository.java:
```diff
@@ -7,11 +7,9 @@
-
-
+    @Query("select  count (t)  from Task t where t.id = :taskId and t.ownerId = :id" )
+    public Integer checkCountByOwnerIdAndTaskId(Long taskId, Long id);
```
можно count (t)>0 как ты предлагал

itananina commented on core-service/src/main/java/com/sanjati/core/services/TimePointService.java:
```diff
@@ -24,21 +25,26 @@
         TimePoint tp;
         if (timePointId != null) {
             tp = timePointRepository.findById(timePointId).orElseThrow(()->new ResourceNotFoundException("Отметка не существует"));
             if (tp.getStatus().equals(TimePointStatus.FINISHED)){
```
тут сравниваем через ==

itananina commented on core-service/src/main/java/com/sanjati/core/services/TimePointService.java:
```diff
@@ -24,21 +25,26 @@
+
+            if(timePointRepository.checkCountTimePoints(taskId)==0){
+
+                taskService.changeStatus(taskId, TaskStatus.ACCEPTED.name());
```
предлагаю в taskService перегрузить метод - добавить сигнатуру с (Long taskId, TaskStatus status), чтобы сервис сам доставал стрингу из енума, а то в глаза бросается

itananina replied:
мб стоит проверять не по наличию таймпоинтов, а просто проверять статус таски - если не в работе, то сделать в работе. мне кажется, менеджер может возвращать задачу из статуса "Утверждается" в статус "Назначена", чтобы сотрудники видели, что чето новое делать надо - вот в этом случае таймпоинты уже будут, а таска не в работе

AleksandrTimonin replied:
менеджер должен сменить статус на назначено, я сделаю доп проверку на статус назначено и тогда наверно не надо будет дёргать каунт

AleksandrTimonin committed:
C-9-1. изменения:
 1) сменил названия в методах репозитория чтобы по названию было понятно что ищем,
 2) сменил критерий поиска чтобы сразу получить буль, хотя сейчас думаю что каунт был бы удобнее
 3) заменил проверку на количество таймпонитов - проверкой на статус "назначено"
 4) сделал пару изменнеий чтобы из базы вынимался сразу статус и от фронта парсился сразу статус

AleksandrTimonin committed:
Merge branch 'lesson-5' into С-9-1

itananina commented on core-service/src/main/java/com/sanjati/core/repositories/TaskRepository.java:
```diff
@@ -2,16 +2,17 @@
-
-
+    @Query("select  count (t) > 0  from Task t where t.id = :taskId and t.ownerId = :id" )
+    public boolean isCountMoreThanZeroByOwnerIdAndTaskId(Long taskId, Long id);
```
нашла штуку без @Query - existsByIdAndOwnerId

itananina commented on core-service/src/main/java/com/sanjati/core/controllers/TaskController.java:
```diff
@@ -109,7 +110,9 @@ public TaskDto getTaskById(@Parameter(description = "ID заявки", required
                                      @PathVariable Long id,
                                  @Parameter(description = "Новый статус заявки", required = true)
-                                     @PathVariable String newStatus,
+
```
не запускал после исправления?  на 118 строчке тут сломалось

itananina replied:
и path сломался

itananina replied:
+ видимо нужен какой-то маппер, чтобы передаваемую в сервис строку преобразовать в енум... короче решение не рабочее совсем

AleksandrTimonin replied:
Запускалось...

itananina commented on core-service/src/main/java/com/sanjati/core/services/TaskService.java:
```diff
@@ -40,7 +40,7 @@ public Task findById(Long id) {
 
     @Transactional
-    public void changeStatus(Long id, String status) {
+    public void changeStatus(Long id, TaskStatus status) {
```
в связи со сменой сигнатуры есть лишние шаги в методе, например на 45 строке поиск енума из строки

AleksandrTimonin committed:
С-9-1. внезапные исправления taskController 118

AleksandrTimonin committed:
С-9-1. внезапные исправления taskController вернул стрингу
